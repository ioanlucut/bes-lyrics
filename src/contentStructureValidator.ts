import { difference, isEmpty, trim, without } from 'lodash-es';
import chalk from 'chalk';
import { COMMA } from '../constants.js';
import { SongSection } from './types.js';
import { isKnownSongSequence } from './utils.js';

export const verifyStructure = (content: string) => {
  const sectionTuples = content
    .split(/(\[.*])/gim)
    .filter(Boolean)
    .map(trim);

  if (!content.includes(SongSection.TITLE)) {
    throw new Error(`${SongSection.TITLE} is missing.`);
  }

  if (!content.includes(SongSection.SEQUENCE)) {
    throw new Error(`${SongSection.SEQUENCE} is missing.`);
  }

  const sequenceIdentifiersFromSections = sectionTuples[3]
    .split(COMMA)
    .map((sequenceChar) => {
      if (!isKnownSongSequence(sequenceChar)) {
        throw new Error(`Unknown "${sequenceChar}" section.`);
      }

      return `[${sequenceChar}]`;
    });

  const sectionsHashMap = {} as Record<string, string>;

  for (
    let sectionIndex = 0;
    sectionIndex < sectionTuples.length;
    sectionIndex = sectionIndex + 2
  ) {
    sectionsHashMap[sectionTuples[sectionIndex] as string] =
      sectionTuples[sectionIndex + 1];
  }

  const mismatchingSequence = difference(
    without(
      Object.keys(sectionsHashMap),
      SongSection.TITLE,
      SongSection.SEQUENCE,
    ),
    sequenceIdentifiersFromSections,
  );
  if (!isEmpty(mismatchingSequence)) {
    throw new Error(
      `The ${chalk.red(
        mismatchingSequence,
      )} tags are present in the content but not in the sequence.`,
    );
  }

  sequenceIdentifiersFromSections.forEach((section) => {
    if (!sectionsHashMap[section]) {
      throw new Error(
        `The ${chalk.red(
          section,
        )} is defined in the sequence but missing as a ${chalk.red(
          section,
        )} section.`,
      );
    }
  });
};
