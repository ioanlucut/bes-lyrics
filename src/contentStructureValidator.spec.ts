import { verifyStructure } from './contentStructureValidator';

const SIMPLE_SONG = `[title]
My custom title

[sequence]
v1,v2,v3,p,p2,p3,c,c2,c3,b,b2,b3

[v1]
Row for v1 

[v2]
Row for v2 

[v3]
Row for v3 

[p]
Row for p 

[p2]
Row for p2 

[p3]
Row for p3 

[c]
Row for c 

[c2]
Row for c2 

[c3]
Row for c3 

[b]
Row for b 

[b2]
Row for b2 

[b3]
Row for b3 

`;

const SONG_WITH_MISMATCHING_SEQUENCE = `[title]
My custom title

[sequence]
v1

[v1]
Row for v1 

[v2]
Row for v2 

[v3]
Row for v3 

[p]
Row for p 

[p2]
Row for p2 

[p3]
Row for p3 

[c]
Row for c 

[c2]
Row for c2 

[c3]
Row for c3 

[b]
Row for b 

[b2]
Row for b2 

[b3]
Row for b3 
`;

const SONG_WITH_MISMATCHING_SECTIONS = `[title]
My custom title

[sequence]
v1,c

[v1]
Row 1`;

describe('contentStructureValidator', () => {
  it('should reject missing title', () => {
    expect(() =>
      verifyStructure(`
My custom title

[sequence]
1,c

[v1]
Row 1`),
    ).toThrowErrorMatchingInlineSnapshot(`"[title] is missing."`);
  });

  it('should reject missing sequence', () => {
    expect(() =>
      verifyStructure(`[title]
My custom title

[v1]
Row 1`),
    ).toThrowErrorMatchingInlineSnapshot(`"[sequence] is missing."`);
  });

  it('should not throw for a correct song', () => {
    expect(() => verifyStructure(SIMPLE_SONG)).not.toThrow();
  });

  it('should throw if the sequence does not contain all the sections', () => {
    expect(() =>
      verifyStructure(SONG_WITH_MISMATCHING_SEQUENCE),
    ).toThrowErrorMatchingInlineSnapshot(
      `"The [31m[v2],[v3],[p],[p2],[p3],[c],[c2],[c3],[b],[b2],[b3][39m tags are present in the content but not in the sequence."`,
    );
  });

  it('should throw if the sections defined in the sequence are not defined', () => {
    expect(() =>
      verifyStructure(SONG_WITH_MISMATCHING_SECTIONS),
    ).toThrowErrorMatchingInlineSnapshot(
      `"The [31m[c][39m is defined in the sequence but missing as a [31m[c][39m section."`,
    );
  });
});
