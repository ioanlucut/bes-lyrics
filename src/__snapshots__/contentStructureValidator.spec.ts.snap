// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contentStructureValidator content should reject correctly for a \`Duplicate section.\`: Duplicate section. 1`] = `"There are duplicates: [c]"`;

exports[`contentStructureValidator content should reject correctly for a \`Missing title!\`: Missing title! 1`] = `"[title] is missing."`;

exports[`contentStructureValidator content should reject correctly for a \`Supported b's chars in the bad order.\`: Supported b's chars in the bad order. 1`] = `"The two sequences/sub-sequences "2" from "b" context and "4" from "b" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Supported c's chars in the bad order.\`: Supported c's chars in the bad order. 1`] = `"The two sequences/sub-sequences "2" from "c" context and "4" from "c" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Supported p's chars in the bad order.\`: Supported p's chars in the bad order. 1`] = `"The two sequences/sub-sequences "2" from "p" context and "4" from "p" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Supported v's chars in the bad order.\`: Supported v's chars in the bad order. 1`] = `"The two sequences/sub-sequences "1" from "v" context and "3" from "v" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Supported v's chars in the bad order.\`: Supported v's chars in the bad order. 2`] = `"The two sequences/sub-sequences "2" from "v" context and "6" from "v" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Supported v's chars in the bad order.\`: Supported v's chars in the bad order. 3`] = `"The two sequences/sub-sequences "3" from "v" context and "9" from "v" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Unsupported chars in sequence.\`: Unsupported chars in sequence. 1`] = `"Unknown "v0" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Unsupported chars in sequence.\`: Unsupported chars in sequence. 2`] = `"Unknown "p1" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Unsupported chars in sequence.\`: Unsupported chars in sequence. 3`] = `"Unknown "b1" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Unsupported chars in sequence.\`: Unsupported chars in sequence. 4`] = `"Unknown "c1" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Unsupported chars in sequence.\`: Unsupported chars in sequence. 5`] = `"Unknown "e1" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (b1 and b1.2 together do not make sense).\`: Wrong sub section order (b1 and b1.2 together do not make sense). 1`] = `"Unknown "b1" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (c1 and c1.2 together do not make sense).\`: Wrong sub section order (c1 and c1.2 together do not make sense). 1`] = `"Unknown "c1" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing b1.2).\`: Wrong sub section order (missing b1.2). 1`] = `"The two sequences/sub-sequences "1" from "1.1" context and "3" from "1.3" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing b2).\`: Wrong sub section order (missing b2). 1`] = `"The two sequences/sub-sequences "2" from "1.2" context and "1" from "v.1" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing c1.2).\`: Wrong sub section order (missing c1.2). 1`] = `"The two sequences/sub-sequences "1" from "1.1" context and "3" from "1.3" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing c2).\`: Wrong sub section order (missing c2). 1`] = `"The two sequences/sub-sequences "2" from "1.2" context and "1" from "v.1" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing p1.2).\`: Wrong sub section order (missing p1.2). 1`] = `"The two sequences/sub-sequences "1" from "1.1" context and "3" from "1.3" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing p2).\`: Wrong sub section order (missing p2). 1`] = `"The two sequences/sub-sequences "2" from "1.2" context and "1" from "v.1" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing v1.2).\`: Wrong sub section order (missing v1.2). 1`] = `"The two sequences/sub-sequences "1" from "1.1" context and "3" from "1.3" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (missing v2).\`: Wrong sub section order (missing v2). 1`] = `"The two sequences/sub-sequences "1" from "1.2" context and "3" from "3.1" context are not consecutive."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (not required to have b1.1).\`: Wrong sub section order (not required to have b1.1). 1`] = `"The "2.1" and "1.1" cannot both end in the same sub sequence."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (not required to have c1.1).\`: Wrong sub section order (not required to have c1.1). 1`] = `"The "2.1" and "1.1" cannot both end in the same sub sequence."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (not required to have p1.1).\`: Wrong sub section order (not required to have p1.1). 1`] = `"The "2.1" and "1.1" cannot both end in the same sub sequence."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (not required to have v1.1).\`: Wrong sub section order (not required to have v1.1). 1`] = `"The "2.1" and "1.1" cannot both end in the same sub sequence."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (p1 and p1.2 together do not make sense).\`: Wrong sub section order (p1 and p1.2 together do not make sense). 1`] = `"Unknown "p1" section."`;

exports[`contentStructureValidator content should reject correctly for a \`Wrong sub section order (v1 and v1.2 together do not make sense).\`: Wrong sub section order (v1 and v1.2 together do not make sense). 1`] = `"The two sequences/sub-sequences "1" from "1" context and "1" from "1.2" context are not consecutive."`;
